services/management/rds.go

```diff
@@ -215,3 +217,143 @@ func (s *DiscoveryService) DiscoverRDS(ctx context.Context, req *managementpb.Di
 	}
 	return nil, err
 }
+
+// AddRDS adds RDS instance.
+func (s *RDSService) AddRDS(ctx context.Context, req *managementpb.AddRDSRequest) (*managementpb.AddRDSResponse, error) {
+	res := new(managementpb.AddRDSResponse)
+
+	if e := s.db.InTransaction(func(tx *reform.TX) error {
+		// tweak according to API docs
+		if req.NodeName == "" {
+			req.NodeName = req.InstanceId
+		}
+		if req.ServiceName == "" {
+			req.ServiceName = req.InstanceId
+		}
+
+		// tweak according to API docs
+		tablestatsGroupTableLimit := req.TablestatsGroupTableLimit
+		if tablestatsGroupTableLimit == 0 {
+			tablestatsGroupTableLimit = defaultTablestatsGroupTableLimit
+		}
+		if tablestatsGroupTableLimit < 0 {
+			tablestatsGroupTableLimit = -1
+		}
+
+		// add RemoteRDS Node
+		node, err := models.CreateNode(tx.Querier, models.RemoteRDSNodeType, &models.CreateNodeParams{
+			NodeName:  req.NodeName,
+			NodeModel: req.NodeModel,
+			AZ:        req.Az,
+			Address:   req.InstanceId,
+			Region:    &req.Region,
+		})
+		if err != nil {
+			return err
+		}
+		invNode, err := services.ToAPINode(node)
+		if err != nil {
+			return err
+		}
+		res.Node = invNode.(*inventorypb.RemoteRDSNode)
+
+		// add RDSExporter Agent
+		if req.RdsExporter {
+			rdsExporter, err := models.CreateAgent(tx.Querier, models.RDSExporterType, &models.CreateAgentParams{
+				PMMAgentID:   models.PMMServerAgentID,
+				NodeID:       node.NodeID,
+				AWSAccessKey: req.AwsAccessKey,
+				AWSSecretKey: req.AwsSecretKey,
+			})
+			if err != nil {
+				return err
+			}
+			invRDSExporter, err := services.ToAPIAgent(tx.Querier, rdsExporter)
+			if err != nil {
+				return err
+			}
+			res.RdsExporter = invRDSExporter.(*inventorypb.RDSExporter)
+		}
+
+		switch req.Engine {
+		case managementpb.DiscoverRDSEngine_DISCOVER_RDS_MYSQL:
```

Comment:
maybe we can extract this duplicated code to somewhere?

Reply:
Most of the duplicate code is already extracted into `models` functions. And the rest is similar, but not exactly a duplicate – there are subtle differences in parameters.
I don't see what we can extract there.